sbin/submitty_autograding_shipper.py

```diff
@@ -355,24 +358,35 @@ def unpack_job(which_machine,which_untrusted,next_directory,next_to_grade):
 
             if not success:
                 return False
-    # archive the results of grading
+
     try:
-        success = packer_unpacker.unpack_grading_results_zip(which_machine,which_untrusted,local_results_zip)
+        with open(local_done_queue_file, 'r') as infile:
+            local_done_queue_obj = json.load(infile)
+        # Check to make certain that the job we received was the correct job
+        if random_identifier != local_done_queue_obj['identifier']:
+            success= False
+            msg = f"{which_machine} returned a stale job (ids don't match). Discarding."
+        else:
+            # archive the results of grading
+            success = packer_unpacker.unpack_grading_results_zip(which_machine,which_untrusted,local_results_zip)
+            msg = "Unpacked job from " + which_machine
     except:
         autograding_utils.log_stack_trace(AUTOGRADING_STACKTRACE_PATH, job_id=JOB_ID, trace=traceback.format_exc())
         autograding_utils.log_message(AUTOGRADING_LOG_PATH, JOB_ID,jobname=item_name,message="ERROR: Exception when unpacking zip. For more details, see traces entry.")
         with contextlib.suppress(FileNotFoundError):
             os.remove(local_results_zip)
+        msg = "ERROR: failure returned from worker machine"
         success = False
 
     with contextlib.suppress(FileNotFoundError):
         os.remove(local_done_queue_file)
 
-    msg = "Unpacked job from " + which_machine if success else "ERROR: failure returned from worker machine"
     print(msg)
     autograding_utils.log_message(AUTOGRADING_LOG_PATH, JOB_ID, jobname=item_name, which_untrusted=which_untrusted, is_batch=is_batch, message=msg)
     return True
 
+def generate_random_string(length):
+  return ''.join(random.choice(string.ascii_uppercase) for i in range(length))
```

Comment:
Let's add this to the submitty_utils python module as we use this same function more or less in three different places.

Reply:
Updated